File: autocontext/src/autocontext/knowledge/soft_hints.py
"""Soft structural hint policy and A/B reporting (AC-796)."""

from __future__ import annotations

from collections import defaultdict
from statistics import fmean
from typing import Any, Literal

HintStyle = Literal["default", "structural", "solution_like"]

STRUCTURAL_HINT_POLICY = (
    "Structural hint policy: prefer constraints, invariants, verification checks, "
    "promising representations, and repair directions; avoid full target solutions, "
    "exact parameter recipes, and route-locking commitments unless the user explicitly asks."
)

_ROUTE_TERMS = ("exact", "set ", "must use", "full solution", "parameter recipe", "route")
_STRUCTURAL_TERMS = ("constraint", "invariant", "check", "verify", "repair", "representation")


def effective_hint_style(*, soft_hints_enabled: bool, hint_style: str) -> HintStyle:
    normalized = hint_style.strip().lower()
    if soft_hints_enabled:
        return "structural"
    if normalized == "structural":
        return "structural"
    if normalized == "solution_like":
        return "solution_like"
    return "default"


def structural_hint_prompt(hint_style: str) -> str:
    return STRUCTURAL_HINT_POLICY if hint_style == "structural" else ""


def build_hint_metadata(text: str, *, hint_style: str, support_evidence: str = "") -> dict[str, Any]:
    lowered = text.lower()
    route_prescriptive = any(term in lowered for term in _ROUTE_TERMS)
    structurally_worded = any(term in lowered for term in _STRUCTURAL_TERMS)
    return {
        "hint_style": hint_style,
        "support_evidence": support_evidence,
        "is_structural": hint_style == "structural" or structurally_worded,
        "route_prescriptive": route_prescriptive,
    }


def build_hint_ab_report(rows: list[dict[str, Any]]) -> dict[str, Any]:
    grouped: dict[str, list[dict[str, Any]]] = defaultdict(list)
    for row in rows:
        grouped[str(row.get("hint_style") or "default")].append(row)
    return {
        "schema_version": 1,
        "styles": {style: _summarize_style(items) for style, items in sorted(grouped.items())},
    }


def _summarize_style(rows: list[dict[str, Any]]) -> dict[str, Any]:
    return {
        "run_count": len(rows),
        "mean_score": _mean(rows, "score"),
        "mean_response_length": _mean(rows, "response_length"),
        "mean_novelty": _mean(rows, "novelty"),
        "rollback_rate": _rate(rows, "rolled_back"),
        "hint_adoption_rate": _rate(rows, "hint_adopted"),
    }


def _mean(rows: list[dict[str, Any]], key: str) -> float | None:
    values = [float(row[key]) for row in rows if isinstance(row.get(key), int | float)]
    return fmean(values) if values else None


def _rate(rows: list[dict[str, Any]], key: str) -> float | None:
    if not rows:
        return None
    return sum(1 for row in rows if bool(row.get(key))) / len(rows)


__all__ = [
    "HintStyle",
    "STRUCTURAL_HINT_POLICY",
    "build_hint_ab_report",
    "build_hint_metadata",
    "effective_hint_style",
    "structural_hint_prompt",
]

File: ts/src/knowledge/soft-hints.ts
export type HintStyle = "default" | "structural" | "solution_like";

export const STRUCTURAL_HINT_POLICY =
  "Structural hint policy: prefer constraints, invariants, verification checks, promising representations, and repair directions; avoid full target solutions, exact parameter recipes, and route-locking commitments unless the user explicitly asks.";

const ROUTE_TERMS = ["exact", "set ", "must use", "full solution", "parameter recipe", "route"];
const STRUCTURAL_TERMS = ["constraint", "invariant", "check", "verify", "repair", "representation"];

export function effectiveHintStyle(softHintsEnabled: boolean, hintStyle: string): HintStyle {
  const normalized = hintStyle.trim().toLowerCase();
  if (softHintsEnabled) return "structural";
  if (normalized === "structural" || normalized === "solution_like") return normalized;
  return "default";
}

export function structuralHintPrompt(hintStyle: string): string {
  return hintStyle === "structural" ? STRUCTURAL_HINT_POLICY : "";
}

export function buildHintMetadata(
  text: string,
  opts: { hintStyle: string; supportEvidence?: string },
): {
  hintStyle: string;
  supportEvidence: string;
  isStructural: boolean;
  routePrescriptive: boolean;
} {
  const lowered = text.toLowerCase();
  const routePrescriptive = ROUTE_TERMS.some((term) => lowered.includes(term));
  const structurallyWorded = STRUCTURAL_TERMS.some((term) => lowered.includes(term));
  return {
    hintStyle: opts.hintStyle,
    supportEvidence: opts.supportEvidence ?? "",
    isStructural: opts.hintStyle === "structural" || structurallyWorded,
    routePrescriptive,
  };
}

export interface HintAbReportRow {
  hintStyle?: string;
  score?: number;
  responseLength?: number;
  novelty?: number;
  rolledBack?: boolean;
  hintAdopted?: boolean;
}

export interface HintAbStyleSummary {
  runCount: number;
  meanScore: number | null;
  meanResponseLength: number | null;
  meanNovelty: number | null;
  rollbackRate: number | null;
  hintAdoptionRate: number | null;
}

export function buildHintAbReport(rows: HintAbReportRow[]): {
  schemaVersion: 1;
  styles: Record<string, HintAbStyleSummary>;
} {
  const grouped = new Map<string, HintAbReportRow[]>();
  for (const row of rows) {
    const style = row.hintStyle ?? "default";
    grouped.set(style, [...(grouped.get(style) ?? []), row]);
  }
  return {
    schemaVersion: 1,
    styles: Object.fromEntries(
      [...grouped.entries()]
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([style, items]) => [style, summarize(items)]),
    ),
  };
}

function summarize(rows: HintAbReportRow[]): HintAbStyleSummary {
  return {
    runCount: rows.length,
    meanScore: mean(rows.map((row) => row.score)),
    meanResponseLength: mean(rows.map((row) => row.responseLength)),
    meanNovelty: mean(rows.map((row) => row.novelty)),
    rollbackRate: rate(rows.map((row) => row.rolledBack)),
    hintAdoptionRate: rate(rows.map((row) => row.hintAdopted)),
  };
}

function mean(values: Array<number | undefined>): number | null {
  const numeric = values.filter(
    (value): value is number => typeof value === "number" && Number.isFinite(value),
  );
  return numeric.length === 0
    ? null
    : numeric.reduce((sum, value) => sum + value, 0) / numeric.length;
}

function rate(values: Array<boolean | undefined>): number | null {
  return values.length === 0 ? null : values.filter(Boolean).length / values.length;
}
